server/e2e_test.go

```diff
@@ -331,35 +331,32 @@ func newTestInbox(conn net.Conn) *testInbox {
 // PopOldest pops the oldest message and returns it or returns an error
 // if no message appears within a short time period.
 func (inbox *testInbox) PopOldest() (*packetInfo, error) {
+	var packet *packetInfo
+
 	err := retry.Do(
 		func() error {
 			inbox.mutex.Lock()
+			defer inbox.mutex.Unlock()
 			if len(inbox.packets) == 0 {
-				inbox.mutex.Unlock()
 				return fmt.Errorf("empty inbox")
 			}
-			// retain the lock if we find a message
+			packet = inbox.packets[0]
+			inbox.packets = inbox.packets[1:]
```

Comment:
the complicated lock/unlock logic made me nervous. I just packed all the action into the retry so the mutex can be dead simple.